### ISSUE TITLE
Dynamic variables "leak" into wrong space when multiple equally named `DynamicVariableSpace` components are duplicated together

### ISSUE THREAD

**Issue by JackTheFoxOtter** [Yellow-Dog-Man/Resonite-Issues#3148]
### Describe the bug?

When you duplicate a hierarchy of multiple nested identically named `DynamicVariableSpace` components, any variables that are directly bound to them will "leak" from the highest dynamic variable space into the other ones. This only happens to named spaces and variables that are directly bound (so prefixed by the space name).

### To Reproduce

Create two nested slots, each with a `DynamicVariableSpace` with the same name, give each a directly bound variable with the same name, but separate values, and duplicate them both together. You'll notice that the value from the parent dynamic variable space will override the value of the child dynamic variable space.

Note how using either `DynamicVariableSpace` components without a name or indirectly bound variables will behave as expected.

### Expected behavior

The two distinct spaces will maintain distinct values in the duplicated hierarchy.

### Screenshots

![Image](https://github.com/user-attachments/assets/ab1a2661-1142-46c9-8917-d0fca50c9ce8)
![Image](https://github.com/user-attachments/assets/ae3f447a-2688-4532-99f1-a1c755a7e21f)

### Resonite Version Number

Beta 2024.10.29.1229

### What Platforms does this occur on?

Windows

### What headset if any do you use?

Valve Index

### Log Files

[J4-C - 2024.10.29.1229 - 2024-10-31 19_57_13.log](https://github.com/user-attachments/files/17592134/J4-C.-.2024.10.29.1229.-.2024-10-31.19_57_13.log)

### Additional Context

I ran into this after putting a lot of work into designing a data structure for a project I'm currently working on. Without this being addressed I won't be able to implement the system I've designed properly.

### Reporters

Myself

**Comment by mpmxyz** [comment]
After glancing at the inner workings I suspect the following chain of events:
0. `inner.CurrentName` and `outer.CurrentName` are null.
1. `outer.OnDuplicate` is called.
2. `outer.CurrentName` is updated.
3. `outer.RelinkAllVariables` is called.
4. Variables within `outer` are linked to `outer`.
5. Variables within `inner` skip `inner` because its CurrentName is still null.
6. Variables within `inner` link with `outer` due to matching name.
7. inner variables will be overwritten with the values from the outer variables.

**Comment by shiftyscales** [comment]
This does sound like it is likely an order of operations issue. From the documentation on the [wiki](https://wiki.resonite.com/Dynamic_Variables):

> "The process of a dynamic variable being associated with a given space is called binding. A dynamic variable component will traverse up the slot hierarchy, including its current slot, looking for an applicable variable space to bind to. If a dynamic variable does not find a dynamic variable space that it can bind to, it will not be accessible outside of the component itself, essentially reducing to a glorified [ValueField](https://wiki.resonite.com/Component:ValueField). A dynamic variable will go through this binding process every time any part of the component changes."

Based on the description of the binding process- it does sound like it should identify / find the child space prior to the parent in this context.

**Comment by JackTheFoxOtter** [comment]
This issue appears to also be part of the underlying problem reported in #3140

**Comment by yoshiyoshyosh** [comment]
> Based on the description of the binding process- it does sound like it should identify / find the child space prior to the parent in this context.

chiming in to say that I wrote the wiki page without knowledge of this edge case. that description of what happens just felt the most intuitive sense to describe. but yeah, given how it's inconsistent with indirect binding/null spaces, definitely feels like a bug

**Comment by JackTheFoxOtter** [comment]
Yeah this is definitely not behaving as it should. It's probably a simple fix, but this bug is currently blocking a big chunk of work I planned to do for the Cloudscape Harvest update, so I'm hoping this can get looked into and addressed some time soon-ish.

**Comment by RustybotPrime** [comment]
Interestingly (and predictably), order offset affects which variables get overwritten.
This has become a blocker for a personal/group project. In our card game, some cards can summon other cards.
Card A can summon card B. Card A has a summon storage slot and a variables slot. If variables has a higher order offset than storage, card A becomes card B when duped. If variables has a lower order offset than storage, card B becomes card A.

**Comment by Mint-Shock** [comment]
I also have made a reproduction item here: https://github.com/Yellow-Dog-Man/Resonite-Issues/issues/5803
++ Video (though in hindsight a bit rambly lol)
